### ISSUE TITLE
Receiving uncompiled source (import/export) with Webpack 2 (inject-loader 2.0.0 / 3.x)

### ISSUE THREAD

**Issue by perry-mitchell** [plasticine/inject-loader#24]
I used inject-loader with webpack 1.x and it functioned perfectly. Porting the code to webpack 2 gives some funny errors.

If I change the code to use `require("inject-loader!...` I get errors like so:
```
ERROR in ./~/inject-loader/lib!./source/ccc/CCCProxy.js
Module parse failed: /home/user/work/my-project/node_modules/inject-loader/lib/index.js!/home/user/work/my-project/node_modules/babel-loader/lib/index.js!/home/user/work/my-project/source/ccc/CCCProxy.js 'import' and 'export' may only appear at the top level (30:0)
You may need an appropriate loader to handle this file type.
|  * @created 2015-08-25
|  */
| import CustomObject from "core/CustomObject";
| import CCCResponse from "ccc/CCCResponse";
| import CCCManagerLocator from "misc/locators/CCCManagerLocator";
 @ ./test/specs/ccc/CCCProxy.spec.js 27:29-66
 @ ./test \.spec\.js$
 @ ./test/index.js
```

If I **leave** it as `"inject!...`, I get other errors:
```
ERROR in ./test/specs/ccc/CCCProxy.spec.js
Module not found: Error: Can't resolve 'inject' in '/home/user/work/my-project/test/specs/ccc'
 @ ./test/specs/ccc/CCCProxy.spec.js 27:29-59
 @ ./test \.spec\.js$
 @ ./test/index.js
```

I'm using Node 6.8 with Webpack 2.1.0-beta.26. I've tried inject-loader 2.0.1 and v3 beta. My code is es6 using import/export being transpiled with babel 6. All code is transpiled correctly to commonjs format which I can confirm by checking the output.

**Comment by TheLarkInn** [comment]
Ahhh what if instead you use `inject-loader!...` instead of just `inject`.

**Comment by perry-mitchell** [comment]
@TheLarkInn That's what I did for my first try 😬 - It **works** (as in webpack recognises the loader), but that's exactly when I get the non-babelified output (with imports etc.). If I look at the output without trying to use inject-loader, all imports/exports are converted to commonjs as expected.

**Comment by rmp135** [comment]
Webpack 2 requires the `-loader` suffix on the loader so you must be `inject-loader`.

What do your webpack, karma and babel configs look like? I would also try rolling inject-loader back to 2.0.1 as I've never gotten v3 beta to work and it installs by default.

**Comment by plasticine** [comment]
@perry-mitchell Thanks for the report, sorry that something seems to be not working correctly.

These errors;

```
ERROR in ./test/specs/ccc/CCCProxy.spec.js
Module not found: Error: Can't resolve 'inject' in '/home/user/work/my-project/test/specs/ccc'
 @ ./test/specs/ccc/CCCProxy.spec.js 27:29-59
 @ ./test \.spec\.js$
 @ ./test/index.js
```

...are as @rmp135 & @TheLarkInn have identified — `inject` alone is not enough for `inject-loader` to be correctly resolved.

As for the actual error — it looks from that error that babel is not compiling es6 imports leading to code like this being output;

```js
// wrapper applied by inject-loader
function injectWrapper() {
  // your code here...
  import foo from 'bar';

  // etc...
} 
```

Which is obviously invalid leading to the error `'import' and 'export' may only appear at the top level`. I suspect that you probably need to change some config to ensure that dependencies are CJS by the time the `inject-loader` wrapper is applied.

Is it possible to provide a failing tests case / example?

**Comment by perry-mitchell** [comment]
> I would also try rolling inject-loader back to 2.0.1 as I've never gotten v3 beta to work and it installs by default.

Thanks @rmp135, I'll try that in the next couple of days when I'm back in front of the workspace.

> inject alone is not enough for inject-loader to be correctly resolved.

Thanks @plasticine - Yeah I guess that's the case there. I'll only be using `inject-loader` from now on, but at least it's possible to alias in the webpack config (or so I've seen).

> I suspect that you probably need to change some config to ensure that dependencies are CJS by the time the inject-loader wrapper is applied.

This is the tricky part.. I'm not sure how to do that. I haven't found any documentation in webpack that covers how to change the processing order when mixing inline + config loaders. I know how to change the order if only in-line or only config, but not together.

**Comment by perry-mitchell** [comment]
This is my config (cobbled together from various files as it's currently being composed per environment):

```
const webpack = require("webpack");

module.exports = {

    entry: {
        test: path.resolve(__dirname, "../test/index.js")
    },

    module: {
        rules: [
            {
                test: /\.html$/,
                use: "raw-loader"
            },
            {
                test: /\.js$/,
                use: "babel-loader",
                exclude: /\/node_modules\//
            },
            {
                test: /\.png$/,
                loaders: [
                    {
                        loader: "url-loader",
                        options: {
                            limit: 8192
                        }
                    },
                    {
                        loader: "image-webpack-loader",
                        query: {
                            progressive: true,
                            bypassOnDebug: true,
                            optimizationLevel: 7,
                            interlaced: false
                        }
                    }
                ]
            },
            {
                test: /\.json$/,
                use: "json-loader"
            }
        ]
    },

    node: {
        setImmediate: false
    },

    output: {
        path: path.resolve(__dirname, "../dist"),
        filename,
        pathinfo: true
    },

    plugins: [
        new webpack.DefinePlugin({
            __ENTRY__: JSON.stringify(scriptType),
            __ENV__: JSON.stringify(environment),
            __DEV__: ["testing", "development"].indexOf(environment) >= 0
        }),
        new webpack.ProvidePlugin({
            setImmediate: "misc/vendor/setimmediate",
            Promise: "promise-polyfill"
        })
    ],

    resolve: {
        extensions: [".js", ".json", ".html"],
        modules: [
            path.resolve(__dirname, "../source"),
            path.resolve(__dirname, "../node_modules"),
            path.resolve(__dirname, "../test")
        ]
    }

};
```

There may be variables not defined but they are in the working environment. Obviously inject-loader is already installed (currently at v3).

**Comment by perry-mitchell** [comment]
I've made 2 examples:

 * [Webpack 2 with inject-loader 2](https://gist.github.com/perry-mitchell/5f9f572f2126b0ca87f914dd074e3ded)
 * [Webpack 2 with inject-loader 3](https://gist.github.com/perry-mitchell/e6838b21e4214a8ab016865c23c62f6c)

Both fail with the same issue.

**Comment by rmp135** [comment]
You need a .babelrc file to actually tell babel how to transform the files.
One to prove it works would be the following.
```javascript
{
  presets: "es2015"
}
```
Also, upgrade to inject-loader@2.0.1. That patch is important, it  won't work otherwise.

**Comment by perry-mitchell** [comment]
Ah I forgot to include that! But I did run them with it. I'll update the examples later. I'll also update inject loader to that ver, thanks.

**Comment by plasticine** [comment]
@perry-mitchell Awesome 🏆 Thanks very much — I’ll have a look at your examples and see if I can work out what’s going on. I would love to make `inject-loader` a bit more developer friendly in this area (at least try and catch some of this stuff and surface it as warnings or something if possible).

Unfortunately due to the nature of the tooling and environment that `inject-loader` works in there will always be many nasty edge cases around config to deal with. :( I’m trying to mess around with making this better for `3.x`.

Also as @rmp135 noted, `2.0.1` is important.

**Comment by perry-mitchell** [comment]
I updated both gists:

 * Inject-loader v2 now uses 2.0.1 and has a .babelrc file
 * Inject-loader v3 now has a .babelrc file

The results? Puzzling. V3 dies with a different error:
![image](https://cloud.githubusercontent.com/assets/3869469/20866336/0a839150-ba34-11e6-81b9-b250fc617680.png)

But... v2 now works!
![image](https://cloud.githubusercontent.com/assets/3869469/20866341/24697102-ba34-11e6-9858-e3b484619cb7.png)

It seems that v2.0.0 and v3.x of inject-loader don't work at all with webpack 2 in this scenario, but 2.0.1 did the trick. The [updated v2 gist](https://gist.github.com/perry-mitchell/5f9f572f2126b0ca87f914dd074e3ded/ae3b3d7c1a4735a1cbbc582ff1211c6e470763ce) now works correctly for me.

**Comment by plasticine** [comment]
@perry-mitchell Awesome — glad it’s working (kinda).

Yep, `2.0.0` was actually completely broken, so I’m glad the version bump got things moving again. :)

`3.x` is still a bit of a work in progress, so I’m not surprised that it’s not working yet. I’m actively working on it though, and it will support Webpack3 once done (hopefully I’ll have time soon).

**Comment by perry-mitchell** [comment]
Thanks for the help @plasticine and others - this is a fantastic tool for testing! I'll be watching the 3.x release closely.

**Comment by plasticine** [comment]
@perry-mitchell I would actually love to use your examples in the repo — would you be open to contributing them? Hopefully it would help others get up to speed. :)

**Comment by perry-mitchell** [comment]
Please go ahead!

I'll be writing a post on this shortly as well, which I'll link here when I'm done 🙂

**Comment by perry-mitchell** [comment]
I'm also happy to make a PR if that's what you meant..

**Comment by plasticine** [comment]
@perry-mitchell Awesome—thanks! Nah, it’s OK, I’ll package it up. Will make sure I credit you though :+1:

**Comment by perry-mitchell** [comment]
Also worth noting to avoid a .babelrc like this one, which I had at work:

```
{
    "presets": [
        ["es2015", { "modules": false }]
    ]
}
```

Disabling the module format change will almost always cause this error to appear.

**Comment by rmp135** [comment]
That's because it isn't converting the modules.

What I settled on was running the tests under a different environment.

```json
{
    "env": {
        "test": {
            "presets": [
                ["es2015"]
            ]
        }
    },
    "presets": [
        ["es2015", { "modules": false }]
    ]
}
```

Then setting `process.env.NODE_ENV = "test";` inside the karma.conf.js file. You could  also set the `NODE_ENV` environment variable on the command line if you wished.

**Comment by plasticine** [comment]
@perry-mitchell @rmp135 Hey peeps, have merged #25 which includes a bunch of fixes — namely webpack2 support (`> 2.2`) and usage examples for both webpack 1 & 2 with babel. Going to release that as `3.0.0-beta3` shortly/now.

**Comment by perry-mitchell** [comment]
Great! I'll give it a go tomorrow or next week in our primary script.
